### DISCUSSION TITLE
How to generate json-crdt-patch by comparing two POJO JSON objects?

### DISCUSSION THREAD

**Discussion by VincentK-ZA** [streamich/json-joy#504] [Q&A] [answered]
Hello

The getting started section for [json-joy/json-crdt](https://jsonjoy.com/libs/json-joy-js/json-crdt/guide/getting-started) shows a basic example where ****specific**/targeted** updates are made to a model, and finally a patch is generated using `model.api.flush()`

```js
// Find "obj" object node at path [].
const obj = model.api.obj([]);

// Overwrite the "counter" last-write-wins register to 25.
// this targets a certain property of the model very specifically
obj.set({ counter: 25 });

// retrieve the JSON CRDT Patch operations which were applied to the model while we were editing it
const patch = model.api.flush();
```

I'm wondering how and if it is possible to generate an array of patch operations to go between two POJO JSON objects?  
Similar to the `compare` function [provided by Fast-JSON-Patch.](https://github.com/Starcounter-Jack/JSON-Patch?tab=readme-ov-file#jsonpatchcomparedocument1-document2-invertible)

Adding the same sample as in the above link here for clarity: 
```js
var documentA = {user: {firstName: "Albert", lastName: "Einstein"}};
var documentB = {user: {firstName: "Albert", lastName: "Collins"}};
var diff = jsonpatch.compare(documentA, documentB);
//diff == [{op: "replace", path: "/user/lastName", value: "Collins"}]
```

My initial though was to use the above compare method from [Fast-JSON-Patch](https://github.com/Starcounter-Jack/JSON-Patch) to get an array of patches, and to then try and build the crdt-patch operations using the [PatchBuilder](https://jsonjoy.com/libs/json-joy-js/json-crdt-patch#The-PatchBuilder-class), but I don't feel like this is the best approach.

The need for this arises from wanting to be able to change the model drastically and freely, but to then send the minimal amount of info across the network for replication. 

Thanks for any assistance and for the ton of work on these libraries!

**Comment by streamich** [comment, marked as answer]
> My initial though was to use the above compare method from [Fast-JSON-Patch](https://github.com/Starcounter-Jack/JSON-Patch) to get an array of patches, and to then try and build the crdt-patch operations using the [PatchBuilder](https://jsonjoy.com/libs/json-joy-js/json-crdt-patch#The-PatchBuilder-class), but I don't feel like this is the best approach.

There is already code which creates [JSON CRDT Patch](https://jsonjoy.com/specs/json-crdt-patch) patches out of [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902) patches, [see](https://github.com/streamich/json-joy/tree/master/src/json-crdt/json-patch).

```ts
import {JsonPatch} from 'json-joy/es2020/json-crdt/json-patch';

const patcher = new JsonPatch(model);
patcher.apply(yourJsonPatch);
```
